### PR TITLE
aws_future

### PR DESCRIPTION
**Pull request by graebm**
`aws_future` was built to [solve the problem of](https://github.com/awslabs/aws-c-common/blob/f77b55d0a48964afc7a5b8bee119d8dd38df7db1/README.md?plain=1#L201-L205): this async function takes a callback, but **sometimes** that callback fires synchronously, and **sometimes** it fires asynchronously

`aws_future` gives the **caller** of an async function control over how the callback is invoked.

This is also an experiment with "C++ templates in C".
Use macros to generate a type-safe API for each type we might store in a future.
For example, to get something like `aws_future<bool>`:
`AWS_DECLARE_FUTURE_T_BY_VALUE(aws_future_bool, bool)`

and now we can write code like:
```
struct aws_future_bool *future = some_async_function();
...
bool result = aws_future_get_result(future);
```

instead of something like:
```
/* this future will contain a bool */
struct aws_future *future = some_async_function();
...
bool result;
aws_future_get_result(future, &result, sizeof(bool));
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by bretambrose — COMMENTED** [review]
Super-off-the-cuff reactions:

Feels way too complex in supporting all of these different possibilities.  Do we really need by-value results? different APIS for pointer, pointer-release, pointer-destructor etc... ownership transfer, non-transfer, etc...?  All the macro-templates?

My suggestion would be a single, simple ref-counted ResultBase structure that you have to use, all results must compose it.

```
struct aws_future_result {
   struct aws_allocator *allocator;
   struct aws_ref_count ref_count;
   void *result;  // points to the containing result
};

...

struct my_result {
   struct aws_future_result base;
   // whatever goes here
}
```

I also feel like the separate callback APIs are perhaps overkill (register, register_event_loop, register_channel). Also register_channel scares me, hate channel tasks.  Maybe instead move the various possibilities into a structure and have a single register_callback API.

Can multiple callbacks be attached to a future (ie a list)?

**Comment by graebm** [comment]
> Can multiple callbacks be attached to a future (ie a list)?

Currently, just 1 callback, but we could support more by storing them in a list. I didn't want to support it without a real-world use-case in C. I know javascript promises let you chain callbacks nicely, but JS does nice things like: don't run 2nd callback if 1st callback raises exception.

> Do we really need by-value results?

My first use of aws_future was to store a bool. It seemed annoying to define a custom `aws_future_bool_result` struct along with with `aws_future_bool_result_new()` and `s_future_bool_result_destroy()`.

I got to this design 1 step at a time

In my [1st iteration](https://github.com/awslabs/aws-c-common/blob/07151af4f90232edc93d0e523e6c4ccf65fca125/include/aws/common/future.h) result was always `void*`, and the "consumer" was required to take ownership of the result and clean it up properly.

Then the first time I tried to use it, the result was a bool, and it seemed really annoying to define a custom `aws_future_bool_result` struct along with with `aws_future_bool_result_new()` and `aws_future_bool_result_destroy()`.

So [my 2nd iteration](https://github.com/awslabs/aws-c-common/blob/3b913ce9f5574467974be17831ff5260d6141e23/include/aws/common/future.h) had a few specific setters/getters for basic by-value types. e.g. `set_bool() / get_bool()` but I expected most use-cases to go through the `set_pointer() / get_pointer()` API.

This was decent. Look [here in aws-c-s3](https://github.com/awslabs/aws-c-s3/blob/5849ecfd63d07a37772b9f39194c3b0b49087a6f/source/s3_auto_ranged_put.c#L786-L787) to see this API in use. You need to document the types stored in each future, and it's OK. But it's not 100% type-safe, and I've seen bugs with mis-use of aws_array_list that have taken years to notice, and I was inspired to try this macro-based-100% type-safe thing.

I'm OK ditching the macros, and just using comments to document the API. But I wanted to experiment with "templates in C via macros", so we could some day develop an `aws_vector<T>` API that is a safer and easier to use wrapper around `aws_array_list`.

**Comment by bretambrose** [comment]
> > Can multiple callbacks be attached to a future (ie a list)?
> 
> Currently, just 1 callback, but we could support more by storing them in a list. I didn't want to support it without a real-world use-case in C. I know javascript promises let you chain callbacks nicely, but JS does nice things like: don't run 2nd callback if 1st callback raises exception.
> 

This wasn't a reference to chaining promises it was a reference to having multiple (sibling) listeners on the result.

File: source/async_stream.c [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    return stream->vtable->read(stream, dest);
+}
+
+/* Data to perform the aws_async_stream_read_to_fill() job */
+struct aws_async_stream_fill_job {
+    struct aws_allocator *alloc;
+    struct aws_async_stream *stream;
+    struct aws_byte_buf *dest;
+    /* Future for each read() step */
+    struct aws_future_bool *read1_future;
```
[the remark this answers is not in the document]
Comment (graebm) [reply source/async_stream.c:55]: in a previous iteration, I called the other function `read_once()`
I'll change it

File: include/aws/io/future.h [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#ifndef AWS_IO_FUTURE_H
+#define AWS_IO_FUTURE_H
+
+/*
+
+// THIS IS AN EXPERIMENTAL AND UNSTABLE API
```
[the remark this answers is not in the document]
Comment (graebm) [reply include/aws/io/future.h:11]: nah

File: source/future.c [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+            void *result = *(void **)result_addr;
+            if (result) {
+                future->result_dtor.destroy(result);
+            }
+        } break;
+
+        case AWS_FUTURE_T_POINTER_WITH_RELEASE: {
+            void *result = *(void **)result_addr;
+            if (result) {
+                future->result_dtor.destroy(result);
```
[the remark this answers is not in the document]
Comment (graebm) [reply source/future.c:75]: 😱

**Review by DmitriyMusatkin — APPROVED** [review]

**Pull request merged**